### ISSUE TITLE
Online/Incremental Training

### ISSUE THREAD

**Issue by abernkopf** [inspirehep/magpie#113]
Hey guys, great tool. This isn't so much an issue as it is a request. Whenever Keras makes the capability easy to use, it'd be great if this tool enabled online/incremental learning. Often the information for this kind of task becomes available a tiny bit at a time, it'd be great if magpie were updated with it when available. 

Anyway, thanks!

**Comment by jstypka** [comment]
hey @abernkopf, thanks for the feedback! You can actually try online learning already with Magpie using the `magpie.batch_train()` function, it's going to load the data from your hard drive and train in batches. Hope that helps!

**Comment by abernkopf** [comment]
Really? I thought `batch_train()` was for when the training data was too big to fit in memory, and thus would be loaded in batches. I poked around a bit and most of the keras community mentioned that online/incremental training wouldn't work very well because of the way that optimizers were structured. 

What I mean to ask is this, so lets say I've trained a model. Sometime later I get some new information, and I want to continue to train the model, but not start from scratch. Some people say that Keras isn't great at this, there seems to be some ambiguity here. `batch_train()` will help me accomplish this, without being detrimental to the model in other aspects?

**Comment by jstypka** [comment]
Ah ok, as you said - the `batch_train()` is for out-of-memory training, I thought this is what you meant.

For the incremental training - you can do it both with Keras and Magpie. Just run `train()` or `batch_train()` few times in a row on the same object and the model will start from where it left off. You can serialize it to disk in the meantime.

**Comment by abernkopf** [comment]
Right, I'd thought this would work, but that there might be some side effects for doing it this way. For example the training rate might get reset, and it might unlearn things it might have previously learned. Do you have any thoughts on how this might work? I suppose it's worth some experimentation.

**Comment by jstypka** [comment]
Regarding the training rate, I'd expect Keras to remember it, but you'd have to ask the folks that develop it about the details. I would imagine that the model might be a bit confused when you throw new data at it after several epochs and unlearn things, but if you feed it the same data I think it should be fine. Trying out is the one way to check! :)

**Comment by lombardi-zz** [comment]
My experience is that the training begins by 0 when i load the model and start a batch_train.

**Comment by ghostviper** [comment]
To online learning，there are many problem can be considered which include corpus cleaning, unbalance data observation and tidy, labels marked， I think it's a Engineering issues not just magpie

**Comment by rajeshkumargp** [comment]
@jstypka  Can you please elaborate ..

""""
For the incremental training - you can do it both with Keras and Magpie. Just run train() or batch_train() few times in a row on the same object and the model will start from where it left off. You can serialize it to disk in the meantime. 
"""" 
batch_train() few times in a row on the same object means... ?

------------------------------------
Please let me know if I am wrong, 
   Assume I have a Magpie model, I need to make it learn for some new training set. 

1. Loading  the existing model,ie existing weights 
2. train with new data such that the initial value of the weights in model is already trained values. 
3. now, the weights get updated and then we will test on new data. 

Am I right .. 
Can you please provide suggestion on this. .. ?

**Comment by jstypka** [comment]
@rajeshkumargp yes, that's pretty much what I meant. You can load weights from one place and then train it on some other data and so on.

**Comment by rajeshkumargp** [comment]
Thanks @jstypka  for  update and clarification.

**Comment by jstypka** [comment]
@rajeshkumargp You compute the word vectors before you run the training, so the only part that is run incrementally is fitting the weights of the neural network. If the model finds a word in text that it's not familiar with, it's going to just ignore it.

**Comment by rajeshkumargp** [comment]
@jstypka  How about Scaler Fitting Step in Incremental Training ?

I have built Word2Vec Word Embedding for preprocessed entire Corpus(same preprocessing done as in Word2Vec Building in Magpie) using Word2vec C Program, which is faster.  

Later I have built Scaler Fitting over the entire corpus. 
I am facing issue as in #168. 

Not sure about Scaler Fitting part in Incremental Training .. 

Can you please guide/clarify on Scaler Fitting part of Incremental Training ?

**Comment by jstypka** [comment]
@rajeshkumargp you want to fit the scaler only once and keep it static for the process of training and testing. You shouldn't fit it on every new batch, then the NN might not be able to learn.

If you your dataset doesn't fit into memory, it's fine to run the scaling only on a smaller subset. It should not have a big impact on performance.

In fact, if you're adventurous you can skip scaling altogether (need to hack Magpie to not complain about it though) and it should work as well. It's going to perform worse and train slower, but it should learn.

**Comment by bigredbug47** [comment]
Hello @jstypka I try incremental training using magpie but it doesn't work as I expected. 

First, my initial data set was had 3 labels: "Food, Drink, Animal":
- Food: 10 rows
- Drink: 10 rows
- Animal: 0 row
After training, I have saved all the model's config files.

Next, I create a small new data set which only have 1 label "Animal" - 3 rows, and load old model, I have train it again and saved the new model. I didn't change anything in the source code.
But when I run predict process by using the previous model, it only returns the label Animal which confident score is >95%.
I thought a model should remember old weight and just update weight when having new information, but it seems to forget all the old data.
What did I do wrong?
Thank a lot.

**Comment by jstypka** [comment]
@bigredbug47 you're right, when training incrementally the model should not forget the old data. I would expect though, that the reason for the problem is how you train it. If I understand correctly, you're training set is 20-25 samples - this is super small. We have been training the model on thousands of samples to get reasonable performance. It depends on the number of epochs and the parameters, but I would imagine that after 20 rows the model just didn't learn anything. Did you perform extensive testing before you ran the incremental training? Was the model any good after 20 samples?
